### DISCUSSION TITLE
Feature request: ability to customize path to .env

### DISCUSSION THREAD

**Discussion by tomdavies** [craftcms/cms#11894] [Ideas] [answered]
Being able to customize the path to `.env` would allow for more flexible project structures, for example when Craft is being used along side other services and it is convenient to share envvars in development via a shared `.env` file.

## Example

This is the (simplified) directory structure that I have for our GH repo template:

```bash
.
├── .env
├── composer.json
├── composer.lock
├── src
│   ├── assets
│   ├── bootstrap.php
│   ├── config
│   ├── craft
│   ├── modules
│   ├── plugins
│   ├── storage
│   ├── templates
│   └── web
└── vendor/

```

Moving .env works completely fine by just editing `bootstrap.php` a little bit:

```php
<?php
/**
 * Shared bootstrap file
 */

// Define path constants
define('CRAFT_BASE_PATH', __DIR__);
define('DOTENV_PATH', dirname(CRAFT_BASE_PATH, 1));
define('CRAFT_COMPOSER_PATH', dirname(CRAFT_BASE_PATH, 1) . '/composer.json');
define('CRAFT_VENDOR_PATH', dirname(CRAFT_BASE_PATH, 1) . '/vendor');

// Load Composer's autoloader
require_once CRAFT_VENDOR_PATH . '/autoload.php';

if (class_exists(Dotenv\Dotenv::class) && file_exists(DOTENV_PATH . '/.env')) {
    // By default, this will allow .env file values to override environment variables
    // with matching names. Use `createUnsafeImmutable` to disable this.
    Dotenv\Dotenv::createUnsafeMutable(DOTENV_PATH)->safeLoad();
}
```

However `craft\services\Config.php::getDotEnvPath()` forces the path to `.env` to `Craft::getAlias('@root/.env')`, and hence the `craft setup` / `craft install` commands won't work as expected (existing values will get read from the root `.env` file, but written to a new one at `src/.env`).

Please, pretty please could we have a `CRAFT_DOTENV_PATH` constant that just _defaults_ to current value?

**Comment by brandonkelly** [comment, marked as answer]
Just added support for a `CRAFT_DOTENV_PATH` constant, for Craft 4.3. (e28fc5d0936243df0669cad1f865532a62b31d51)

Alternatively, you can point `CRAFT_BASE_PATH` to the top level directory, and override the `config/`, `storage` and `templates/` directory path constants (rather than `CRAFT_COMPOSER_PATH` and `CRAFT_VENDOR_PATH`):

```php
$basePath = __DIR__;

// Define path constants
define('CRAFT_BASE_PATH', dirname($basePath));
define('CRAFT_CONFIG_PATH', "$basePath/config");
define('CRAFT_STORAGE_PATH', "$basePath/storage");
define('CRAFT_TEMPLATES_PATH', "$basePath/templates");
```

**Reply by tomdavies** [reply]
🙏 thanks @brandonkelly, good tip re `CRAFT_BASE_PATH` too

**Reply by brandonkelly** [reply]
Craft 4.3 was released earlier this week with this change!
